Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -2,119 +2,300 @@
+require "azure/storage"
+require 'json' # for registry content
+require "securerandom" # for generating uuid.
 
-# Reads events from Azure Blobs
-class LogStash::Inputs::Azureblob < LogStash::Inputs::Base
+# Registry item to coordinate between mulitple clients
+class LogStash::Inputs::RegistryItem
+  attr_accessor :file_path, :etag, :offset, :reader
 
+  # Allow json serialization.
+  def as_json(options={})
+      {
+          file_path: @file_path,
+          etag: @etag,
+          reader: @reader,
+          offset: @offset
+      }
+  end # as_json
+
+  def to_json(*options)
+      as_json(*options).to_json(*options)
+  end # to_json
+
+  def initialize(file_path, etag, reader, offset = 0)
+      @file_path = file_path
+      @etag = etag
+      @reader = reader
+      @offset = offset
+  end # initialize
+end # class RegistryItem
+
+
+# Logstash input plugin for Azure Blobs
+#
+# This logstash plugin gathers data from Microsoft Azure Blobs
+class LogStash::Inputs::LogstashInputAzureblob < LogStash::Inputs::Base
   config_name "azureblob"
-  milestone 1
-  
+
+  # If undefined, Logstash will complain, even if codec is unused.
   default :codec, "json_lines"
-  
+
+  # Set the account name for the azure storage account.
   config :storage_account_name, :validate => :string
+  
+  # Set the key to access the storage account.
   config :storage_access_key, :validate => :string
   
+  # Set the container of the blobs.
   config :container, :validate => :string
-  config :sleep_time, :validate => :number, :default => 10
-  config :endpoint, :validate => :string, :default => "core.windows.net"
-  config :backupmode, :validate => :boolean, :default => false 
   
-  def initialize(*args)
-    super(*args)
-  end # def initialize
+  # Set the endpoint for the blobs.
+  #
+  # The default, `core.windows.net` targets the public azure.
+  config :endpoint, :validate => :string, :default => 'core.windows.net'
+
+  # Set the value of using backup mode.
+  config :backupmode, :validate => :boolean, :default => false, :deprecated => true, :obsolete => 'This option is obsoleted and the settings will be ignored.'
+  
+  # Set the value for the registry file.
+  #
+  # The default, `data/registry`, is used to coordinate readings for various instances of the clients.
+  config :registry_path, :validate => :string, :default => 'data/registry'
   
+  # Set how frequently messages should be sent.
+  #
+  # The default, `30`, means trigger a reading for the log every 30 seconds.
+  config :interval, :validate => :number, :default => 30
+
   public
   def register
-    Azure.configure do |config|
-      config.storage_account_name = @storage_account_name
-      config.storage_access_key = @storage_access_key
-      config.storage_blob_host = "https://#{@storage_account_name}.blob.#{@endpoint}"
-    end
-    @azure_blob = Azure::Blob::BlobService.new
+    # this is the reader # for this specific instance.
+    @reader = SecureRandom.uuid
```

Comment:
I actually think logstash can crash more often than you think. Also, it's possible that your update to the registry can fail due to network or other reasons. For approach 2, you already handle the case where multiple clients trying to create the registry files, so I don't see how that would be different compared to this case. Your approach 3 doesn't scale, and your approach 4 is not trivial.

Looking at all these options again, I kind of like option 2. It's not really much harder to implement. As for feedback, we can gather feedback first, but I am afraid a lot of users run into this problem and it's not easy to discover. ie. It may take you a few days before finding that one of your log file is not transferring.